File: 03-Frontend/welps/src/app/common/puppy.ts
import { AnimalCharacter } from "./animal-character.enum";
import { Bread } from "./bread";



export class Puppy {
    constructor(
      private _id?: number,
      private _puppyId?: number,
      private _name?: string,
      private _image?: string,
      private _height?: number,
      private _weight?: number,
      private _color?: string,
      private _characteristic?: AnimalCharacter[],
      private _race?: Bread,
      private _price?: number,
    ) {}
  
    
    get id(): number | undefined {  return this._id;            }   
    set id(value: number | undefined) {   this._id = value;     }  

    get puppyId(): number | undefined { return this._puppyId;   }
    set puppyId(puppyId: number | undefined){   this._puppyId = puppyId;}
  
    get name(): string | undefined{   return this._name;        }
    set name(name: string | undefined){   this._name = name;    }
    
    get weight(): number | undefined {  return this._weight;    }
    set weight(value: number | undefined) {this._weight = value;}
  
    get price(): number | undefined {   return this._price;     }
    set price(value: number | undefined) {this._price = value;  }
  
    get image(): string | undefined {   return this._image;     }
    set image(value: string | undefined) {this._image = value;  }
  
    
    get color(): string | undefined {     return this._color;     }
    set color(color: string){   this._color = color;              }

    get height(): number | undefined {  return this._height;      }
    set height(height:number){          this._height = height;    }

    get race(): Bread | undefined {    return this._race;          }
    set race(race: Bread ){     this._race = race;                  }

    get characteristic(): AnimalCharacter[] | undefined{ return this._characteristic  }
    set characteristic(ch: AnimalCharacter[]){   this._characteristic = ch;           }

    toJSON(): object {
      return {
          id: this._id,
          puppyId: this._puppyId,
          name: this._name,
          image: this._image,
          height: this._height,
          weight: this._weight,
          color: this._color,
          characteristic: this._characteristic,
          race: this._race,
          price: this._price,
      };
  }
  }

File: 02-Backend/Puppies/src/main/java/de/irahi/welp/entity/Puppy.java
package de.irahi.welp.entity;

import java.math.BigDecimal;
import java.util.Objects;
import java.util.Set;

import de.irahi.welp.component.AnimalCharacterConverter;
import de.irahi.welp.props.AnimalCharacter;
import jakarta.persistence.Access;
import jakarta.persistence.AccessType;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import lombok.Data;

@Entity
@Data
@Access(AccessType.FIELD)
public class Puppy extends Product{
	
	@Column(name="weight")
	private BigDecimal weight;
	@Column(name="color")
	private String color;
	@Column(name="height")
	private BigDecimal height;
	
	
	@Column(name="characteristic", columnDefinition = "VARCHAR")
	@Convert(converter= AnimalCharacterConverter.class)
	private Set<AnimalCharacter> characteristic;
	
	
	@ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
	@JoinColumn(name="race_id", referencedColumnName = "id")
	private Breed race;
	
    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;
	
	@Override
	public int hashCode() {
		return Objects.hash(super.hashCode(), weight, color, height, characteristic, race);
	}
	

}
